`Tutorial groups`: Improve functionality of import button

az108 opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check all tasks by putting an x in the [ ] (don't: [x ], [ x], do: [x]). Remove not applicable tasks and do not leave them unchecked -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- Remove tasks that are not applicable for your PR. Please only put the PR into ready for review, if all relevant tasks are checked! -->
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Client
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I documented the TypeScript code using JSDoc style.

### Motivation and Context
This PR is a follow up to the issue mentioned in: https://github.com/ls1intum/Artemis/pull/8710
where the description of importing tutorial groups was adjusted and improved. Now the aim is to improve the import so that more headers are accepted in a CSV and more information can be extracted and used.

### Description
I added a refined explanation on how to create the csv file to import tutorial groups and added the new headers to it aswell. I additionally added the example table 4 on how to create a csv file with optional headers. On server aswell as client side i adjusted the import DTO allowing more fields to be imported and used.

Additionally I implemented the requested changes from my previous PRs in this follow up PR. This included adjusting the import text so that the headers mentioned in the examples are also listed above on the left side and changing the export format for the export to JSON functionality so that students get displayed in a list.

Before:
![image](https://github.com/ls1intum/Artemis/assets/92953467/c5a5c62a-fab3-4435-9435-721b499ff510)
After:
<img width="1022" alt="image" src="https://github.com/ls1intum/Artemis/assets/92953467/0de15021-6c49-49f3-826b-0353d70efa04">
![image](https://github.com/ls1intum/Artemis/assets/92953467/116806ef-7c92-4161-b995-ff445523a479)

### Steps for Testing
- **Click on some Course**: Click on a Course where you have instructor rights
- **Navigate to Manage**: Click on the Manage Icon 
- **Navigate to Tutorials**: Click on the Manage Section 
- **Enable Tutorials**: Provide a basic configuration by entering dates
- **Try to Import**: Use the button and verify the new details 

Please make sure that the file you use is delimited properly with commas. 
For example: if you have the headers *Title,MatriculationNumber* then you can not just have a row with a title. You need to have a row which at least also sets a empty Matriculation Number. This would mean a row for these headers could look like folliowing: *exampleGroup,*  

This is how you could test headers but leave them empty in the rows.


### Testserver States
> [!NOTE]
> These badges show the state of the test servers.
> Green = Currently available, Red = Currently locked

![](https://byob.yarr.is/ls1intum/Artemis/artemis-test1)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test2)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test3)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test4)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test5)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test6)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test9)

### Review Progress
<!-- Each PR should be reviewed by at least two other developers. The code, the functionality (= manual test) and the exam mode need to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->
<!-- All PRs that might affect the exam mode (e.g. change a client component that is also used in the exam mode) need an additional verification that the exam mode still works. -->

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- The line coverage must be above 90% for changes files and you must use extensive and useful assertions for server tests and expect statements for client tests. -->
<!-- Note: Use the table below and confirm in the last column that you have implemented extensive assertions for server tests and expect statements for client tests. -->
<!--       You can use `supporting_script/generate_code_cov_table/generate_code_cov_table.py` to automatically generate one from the corresponding Bamboo build plan artefacts. -->
<!--       Remove rows with only trivial changes from the table. -->

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [CsvDownloadService](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6711-10/artifact/shared/Coverage-Report-Client-Tests/app/shared/util/index.html) | 100%  | ✅ |
| [tutorial-groups-registration-import-dto.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6711-10/artifact/shared/Coverage-Report-Client-Tests/app/entities/tutorial-group/index.html) | 100% | ✅ |
[tutorial-groups-registration-import-dialog.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6677-10/artifact/shared/Coverage-Report-Client-Tests/app/course/tutorial-groups/tutorial-groups-management/tutorial-groups/tutorial-groups-management/tutorial-groups-import-dialog/index.html) | 93.52% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [TutorialGroupService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6711-10/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.tutorialgroups/index.html) | 94% | ✅ |
| [TutorialGroupResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6711-10/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.tutorialgroups/index.html) | 92% | ✅ |

JohannesStoehr commented on src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:
```diff
@@ -543,14 +543,16 @@ public enum TutorialGroupImportErrors {
      */
     @JsonInclude(JsonInclude.Include.NON_EMPTY)
     public record TutorialGroupRegistrationImportDTO(@Nullable String title, @Nullable StudentDTO student, @Nullable Boolean importSuccessful,
```
Is there a reason why you do not include the tutor login or the room?

az108 replied:
The Tutor Login was not included because I found it unintuitive to match tutors via CSV and would break the structure of the CSV format which is created by my other PR. I have adjusted the import process so that exporting tutor groups from the previous year can be directly used to import for the current year. While creating a tutor group, it makes more sense, in my opinion, to use tutors available in an Artemis course, which changes every year. Therefore, I designed it so that you can import all the fields and then select a tutor within Artemis. Initially, the room was planned to be imported and is included in the export. However, Artemis has a tutorial group design where a room can only be added if you create a tutorial group schedule, which is optional and requires fixed dates. This was causing additional problems, so addressing it should be part of a follow-up PR, as it would break the scope of this one.

SimonEntholzer reviewed (changes requested):
![image](https://github.com/ls1intum/Artemis/assets/33342534/26a9fa45-9d83-48dc-ac57-1c6a6d3f5aab)

I like the changes and the code looks good. Tested on TS3, and I found the import doesn't work. Am I doing something wrong here/forget?

az108 commented:
Thanks for pointing this edge case out. There was a small error in the validation where if you would not import a specific field then a value would be set for a new tutorial group which is invalid. I fixed this issue @SimonEntholzer  👍

beyzaaltuntas reviewed (changes requested):
Hi, 
I tested this changes on TS-3. I clicked on download example .csv buttons, and chose them from file uploader. When I tried to parse them, it didn't work. I am using macOS. Am I doing something wrong?

florian-glombik commented on src/main/webapp/app/course/tutorial-groups/tutorial-groups-management/tutorial-groups/tutorial-groups-management/tutorial-groups-import-dialog/tutorial-groups-registration-import-dialog.component.html:
```diff
@@ -1,42 +1,203 @@
+                        </tr>
+                    </thead>
+                    <tbody>
+                        <tr>
+                            <th jhiTranslate="artemisApp.tutorialGroupImportDialog.table.groupA"></th>
+                            <td>123456</td>
+                            <td>John</td>
+                            <td>Doe</td>
+                            <td>Main Campus</td>
+                            <td>English</td>
+                            <td>Extra info</td>
+                            <td>20</td>
+                            <td>true</td>
+                        </tr>
+                        <tr>
+                            <th jhiTranslate="artemisApp.tutorialGroupImportDialog.table.groupA"></th>
+                            <td>555555</td>
+                            <td>Jonas</td>
+                            <td>Krapfen</td>
+                            <td>Main Campus</td>
+                            <td>English</td>
+                            <td>Extra info</td>
+                            <td>20</td>
+                            <td>true</td>
+                        </tr>
+                        <tr>
+                            <th jhiTranslate="artemisApp.tutorialGroupImportDialog.table.groupB"></th>
+                            <td>789012</td>
+                            <td>Anna</td>
+                            <td>Smith</td>
+                            <td>City Campus</td>
+                            <td>German</td>
+                            <td>Extra info</td>
+                            <td>15</td>
+                            <td>false</td>
+                        </tr>
```
I think that data driven approaches are usually better (less error prone due to types and no repetition in the HTML), however, for this simple static information I do not have a strong opinion - does not need to be changed

az108 replied:
not necessary

florian-glombik commented on src/main/webapp/app/course/tutorial-groups/tutorial-groups-management/tutorial-groups/tutorial-groups-management/tutorial-groups-import-dialog/tutorial-groups-registration-import-dialog.component.ts:
```diff
@@ -232,6 +243,11 @@ export class TutorialGroupsRegistrationImportDialogComponent implements OnInit,
                     lastName: csvRow[usedLastNameHeader]?.trim() || '',
                 } as StudentDTO;
+                registration.campus = csvRow[usedCampusHeader]?.trim() || '';
+                registration.capacity = csvRow[usedCapacityHeader] ? parseInt(csvRow[usedCapacityHeader], 10) : undefined;
```
Please use the Number constructor instead, or is there a particular reason why we cannot use it?

```suggestion
                registration.capacity = csvRow[usedCapacityHeader] ? Number(csvRow[usedCapacityHeader]) : undefined;
```

az108 replied:
done 👍

florian-glombik commented on src/main/webapp/app/entities/tutorial-group/tutorial-group-import-dto.model.ts:
```diff
@@ -1,8 +1,14 @@
-    public student: StudentDTO;
+    title: string;
+    student: StudentDTO;
+    campus?: string;
+    capacity?: number;
+    language?: string;
+    additionalInformation?: string;
+    isOnline?: string;
```
I stumbled across this aswell, why is `isOnline` a string?

az108 replied:
Parsing a String in an CSV worked out. Parsing a boolean lead to errors. Thats why

florian-glombik commented on src/main/webapp/i18n/de/tutorialGroupImport.json:
```diff
@@ -6,72 +6,89 @@
             "explanations": {
-                "description": "Dieser Dialog kann verwendet werden, um Übungsgruppen und deren Registrierungen aus einer CSV-Datei zu importieren. Dabei werden Übungsgruppen erstellt, wenn sie noch nicht existieren. Die Studierende werden von ihrer aktuellen Übungsgruppe abgemeldet und in der neuen Übungsgruppe registriert.",
-                "requirements": "Jede Zeile muss immer einen Wert für den Übungsgruppen Titel enthalten. Wenn nicht nur Übungsgruppen erstellt werden sollen, sondern auch die Registrierungen von Studierenden, muss die Zeile einen Wert für die Immatrikulationsnummer oder für den Benutzernamen enthalten. Dies ist erforderlich, um den Studierenden zu identifizieren. Werte für den Nachnamen und Vorna ...
+                "description": "Dieses Dialogfeld ermöglicht es, Tutoriumsgruppen und Studierendenregistrierungen aus einer CSV-Datei zu importieren. Wenn eine Tutoriumsgruppe noch nicht existiert, wird sie erstellt. Wenn Studierende in der CSV-Datei aufgeführt sind, werden sie von ihrer aktuellen Tutoriumsgruppe in die neue, in der CSV-Datei angegebene, Gruppe verschoben. Falls eine Tutoriumsgru ...
```
```suggestion
                "description": "Dieses Dialogfeld ermöglicht es, Tutoriumsgruppen und Studierendenregistrierungen aus einer CSV-Datei zu importieren. Wenn eine Tutoriumsgruppe noch nicht existiert, wird sie erstellt. Wenn Studierende in der CSV-Datei aufgeführt sind, werden sie von ihrer aktuellen Tutoriumsgruppe in die neue, in der CSV-Datei angegebene, Gruppe verschoben. Falls eine Tutoriumsgruppe schon existiert, aber keine Studierenden registriert wird diese nicht importiert.",
```

az108 replied:
done 👍

florian-glombik commented on src/main/webapp/i18n/de/tutorialGroupImport.json:
```diff
@@ -6,72 +6,89 @@
+        "errorMessages": {
+            "genericErrorMessage": "Import von Übungsgruppen sind fehlgeschlagen!",
+            "withoutTitle": "Die folgenden Zeilen enthalten keinen Übungsgruppen Titel: ",
+            "invalidTitle": "Die folgenden Zeilen haben einen ungültigen Titel für die Übungsgruppe (Titel dürfen nur Buchstaben, Zahlen, Leerzeichen und Bindestriche und maximal 19 Zeichen enthalten): ",
```
```suggestion
            "invalidTitle": "Die folgenden Zeilen haben einen ungültigen Titel für die Übungsgruppe (Titel dürfen nur Buchstaben, Zahlen, Leerzeichen, Bindestriche und maximal 19 Zeichen enthalten): ",
```

az108 replied:
done 👍

florian-glombik reviewed (commented):
This is a lot of descriptive text, I think it would be better to display it as a table
e.g. `key before parsing` and `after parsing In CSV`

<img width="638" alt="image" src="https://github.com/ls1intum/Artemis/assets/63976129/f42c7f33-a80d-400f-a72a-c5ddfd7d0450">

az108 commented:
@beyzaaltuntas Did you also fill out the CSV with information about tutorial groups you want to import?

sarpsahinalp reviewed (commented):
I downloaded the example with minimum required headers to parse a CSV. When I tried to import it I got the following error:
![Screenshot 2024-06-15 161443](https://github.com/ls1intum/Artemis/assets/92453703/a6c192d3-586c-4941-9455-e06ea33756c6)

az108 commented on src/main/webapp/app/course/tutorial-groups/tutorial-groups-management/tutorial-groups/tutorial-groups-management/tutorial-groups-import-dialog/tutorial-groups-registration-import-dialog.component.html:
```diff
@@ -1,42 +1,203 @@
+                <li jhiTranslate="artemisApp.tutorialGroupImportDialog.supportedHeaderNames.alwaysRequired"></li>
                 <ul>
-                    <li>{{ 'artemisApp.tutorialGroupImportDialog.supportedHeaderNames.tutorialGroupTitle' | artemisTranslate: { headers: supportedTitleHeader } }}</li>
+                    <li jhiTranslate="artemisApp.tutorialGroupImportDialog.supportedHeaderNames.tutorialGroupTitle"></li>
```
i removed the parameters from the translate file

az108 commented on src/main/webapp/app/course/tutorial-groups/tutorial-groups-management/tutorial-groups/tutorial-groups-management/tutorial-groups-import-dialog/tutorial-groups-registration-import-dialog.component.ts:
```diff
@@ -206,14 +211,20 @@ export class TutorialGroupsRegistrationImportDialogComponent implements OnInit,
         }
         // get the used headers from the first csv row object returned by the parser
-        const parsedHeaders = Object.keys(csvRows.first() || []);
+        let parsedHeaders = Object.keys(csvRows[0] || []);
```
done 👍

az108 commented on src/main/webapp/app/course/tutorial-groups/tutorial-groups-management/tutorial-groups/tutorial-groups-management/tutorial-groups-import-dialog/tutorial-groups-registration-import-dialog.component.ts:
```diff
@@ -246,6 +262,11 @@ export class TutorialGroupsRegistrationImportDialogComponent implements OnInit,
         }
     }
 
+    // Function to remove all spaces in header names
+    removeSpacesFromHeaderName(header: string) {
```
done 👍

az108 commented on src/main/webapp/app/course/tutorial-groups/tutorial-groups-management/tutorial-groups/tutorial-groups-management/tutorial-groups-import-dialog/tutorial-groups-registration-import-dialog.component.ts:
```diff
@@ -259,6 +280,33 @@ export class TutorialGroupsRegistrationImportDialogComponent implements OnInit,
         });
     }
 
+    generateCSV(example: number) {
+        let csvContent: string;
+        switch (example) {
+            case 1:
+                csvContent = 'data:text/csv;charset=utf-8,Title\n';
+                break;
+            case 2:
+                csvContent = 'data:text/csv;charset=utf-8,Title,Matriculation Number,First Name,Last Name\n';
+                break;
+            case 3:
+                csvContent = 'data:text/csv;charset=utf-8,Title,Login,First Name,Last Name\n';
+                break;
+            case 4:
+                csvContent = 'data:text/csv;charset=utf-8,Title,Registration Number,First Name,Last Name,Campus,Language,Additional Information,Capacity,Is Online\n';
```
done 👍

az108 commented on src/main/webapp/app/course/tutorial-groups/tutorial-groups-management/tutorial-groups/tutorial-groups-management/tutorial-groups-import-dialog/tutorial-groups-registration-import-dialog.component.ts:
```diff
@@ -259,6 +280,33 @@ export class TutorialGroupsRegistrationImportDialogComponent implements OnInit,
+            default:
+                csvContent = '';
+        }
+        const encodedUri = encodeURI(csvContent);
+        const link = document.createElement('a');
+        link.setAttribute('href', encodedUri);
+        link.setAttribute('download', `example${example}.csv`);
+        document.body.appendChild(link);
+        link.click();
+        document.body.removeChild(link);
```
done 👍

az108 commented:
@florian-glombik after turning the above text also to a table, the import page just didn't look good anymore with 6 to 7 tables. I adjusted the problem you mentioned and extracted the similar text tho and now it should be less text 👍

JohannesWt reviewed (dismissed):
Tested in Testing Session on TS4 - works as promised 👍

JohannesStoehr reviewed (dismissed):
Code looks good to me.
Also tested in Testing session

BBesrour reviewed (dismissed):
Demonstrated in a testing session and tested on TS4, works as expected

EneaGore reviewed (dismissed):
Tested on TS4. Works as expected.

coolchock reviewed (dismissed):
tested in testing session, works as expected

JohannesStoehr reviewed (approved):
Reapprove

EneaGore reviewed (approved):
reapprove

BBesrour reviewed (approved):
Reapprove

coolchock reviewed (approved):
re-approve
